safehttp/header_test.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestSet(t *testing.T) {
+	h := newHeader(http.Header{})
+	if err := h.Set("Foo-Key", "Bar-Value"); err != nil {
+		t.Fatalf(`h.Set("Foo-Key", "Bar-Value") got err: %v want: nil`, err)
+	}
+	if got, want := h.Get("Foo-Key"), "Bar-Value"; got != want {
+		t.Errorf(`h.Get("Foo-Key") got: %q want %q`, got, want)
+	}
+}
+
+func TestSetCanonicalization(t *testing.T) {
+	h := newHeader(http.Header{})
+	if err := h.Set("fOo-KeY", "Bar-Value"); err != nil {
+		t.Fatalf(`h.Set("fOo-KeY", "Bar-Value") got err: %v want: nil`, err)
+	}
+	if got, want := h.Get("FoO-kEy"), "Bar-Value"; got != want {
+		t.Errorf(`h.Get("FoO-kEy") got: %q want %q`, got, want)
+	}
+}
+
+func TestSetSetCookie(t *testing.T) {
+	h := newHeader(http.Header{})
+	err := h.Set("Set-Cookie", "x=y")
+	if got, want := err.Error(), "can't write to Set-Cookie header"; got != want {
+		t.Errorf(`h.Set("Set-Cookie", "x=y") got: %v want: %v`, got, want)
+	}
+	if diff := cmp.Diff([]string{}, h.Values("Set-Cookie")); diff != "" {
+		t.Errorf(`h.Values("Set-Cookie") mismatch (-want +got):\n%s`, diff)
+	}
+}
+
+func TestSetImmutable(t *testing.T) {
+	h := newHeader(http.Header{})
+	h.MarkImmutable("Foo-Key")
+	err := h.Set("Foo-Key", "Bar-Value")
+	if got, want := err.Error(), "immutable header"; got != want {
+		t.Errorf(`h.Set("Foo-Key", "Bar-Value") got: %v want: %v`, got, want)
+	}
+	if diff := cmp.Diff([]string{}, h.Values("Foo-Key")); diff != "" {
+		t.Errorf(`h.Values("Foo-Key") mismatch (-want +got):\n%s`, diff)
+	}
+}
+
+func TestAdd(t *testing.T) {
```

Comment:
Add a test that checks the header values for the same key are **not** being magically sorted anywhere.

Additionally, add a test that checks the behavior of adding multiple equal key-value pairs.